File: backend/dormammu/agent/profiles.py
from __future__ import annotations

from collections.abc import Sequence
from dataclasses import dataclass, field, replace
from pathlib import Path
from typing import TYPE_CHECKING, Any, Mapping

from dormammu.agent.permissions import (
    AgentPermissionPolicy,
    WorktreePermissionPolicy,
    merge_permission_policy,
)
from dormammu.agent.role_config import ROLE_NAMES, AgentsConfig, RoleAgentConfig
from dormammu.agent.role_taxonomy import ROLE_TAXONOMY

if TYPE_CHECKING:
    from dormammu.agent.manifest_loader import LoadedAgentDefinition

BUILTIN_PROFILE_SOURCE = "built_in"
CONFIGURED_PROFILE_SOURCE = "configured"
PROJECT_PROFILE_SOURCE = "project"
USER_PROFILE_SOURCE = "user"
PROFILE_RUNTIME_METADATA_KEY = "dormammu_runtime"


@dataclass(frozen=True, slots=True)
class AgentProfile:
    """Typed internal profile for role-backed agent execution."""

    name: str
    description: str
    source: str = BUILTIN_PROFILE_SOURCE
    prompt_body: str | None = None
    cli_override: Path | None = None
    model_override: str | None = None
    permission_policy: AgentPermissionPolicy = field(default_factory=AgentPermissionPolicy)
    preloaded_skills: tuple[str, ...] = ()
    metadata: dict[str, Any] = field(default_factory=dict)

    def resolve_cli(self, active_agent_cli: Path | None) -> Path | None:
        return self.cli_override if self.cli_override is not None else active_agent_cli

    @property
    def worktree_policy(self) -> WorktreePermissionPolicy:
        return self.permission_policy.worktree

    def to_dict(self) -> dict[str, object]:
        return {
            "name": self.name,
            "description": self.description,
            "source": self.source,
            "prompt_body": self.prompt_body,
            "cli_override": str(self.cli_override) if self.cli_override is not None else None,
            "model_override": self.model_override,
            "permission_policy": self.permission_policy.to_dict(),
            "preloaded_skills": list(self.preloaded_skills),
            "metadata": dict(self.metadata),
        }


_ROLE_DESCRIPTIONS: dict[str, str] = {
    entry.name: entry.description for entry in ROLE_TAXONOMY
}

ROLE_TO_PROFILE_NAME: dict[str, str] = {role: role for role in ROLE_NAMES}


def built_in_permission_policy_for_role(role: str) -> AgentPermissionPolicy:
    if role not in ROLE_TO_PROFILE_NAME:
        raise ValueError(f"Unknown role: {role!r}. Valid roles: {ROLE_NAMES}")
    return AgentPermissionPolicy()


BUILTIN_AGENT_PROFILES: tuple[AgentProfile, ...] = tuple(
    AgentProfile(
        name=ROLE_TO_PROFILE_NAME[role],
        description=_ROLE_DESCRIPTIONS[role],
        permission_policy=built_in_permission_policy_for_role(role),
    )
    for role in ROLE_NAMES
)

_BUILTIN_PROFILES_BY_NAME: dict[str, AgentProfile] = {
    profile.name: profile for profile in BUILTIN_AGENT_PROFILES
}


def _role_override_present(role_config: RoleAgentConfig | None) -> bool:
    return role_config is not None and (
        role_config.cli is not None
        or role_config.model is not None
        or role_config.permission_policy is not None
    )


def profile_name_for_role(
    role: str,
    role_config: RoleAgentConfig | None = None,
) -> str:
    if role not in ROLE_TO_PROFILE_NAME:
        raise ValueError(f"Unknown role: {role!r}. Valid roles: {ROLE_NAMES}")
    if role_config is not None and role_config.profile is not None:
        return role_config.profile
    return ROLE_TO_PROFILE_NAME[role]


def built_in_profiles() -> tuple[AgentProfile, ...]:
    return BUILTIN_AGENT_PROFILES


def built_in_profile_for_role(role: str) -> AgentProfile:
    profile_name = profile_name_for_role(role)
    profile = _BUILTIN_PROFILES_BY_NAME.get(profile_name)
    if profile is None:
        raise ValueError(
            f"Role {role!r} maps to profile {profile_name!r}, "
            "but no built-in profile with that name is available."
        )
    return profile


def is_built_in_profile_name(profile_name: str) -> bool:
    return profile_name in _BUILTIN_PROFILES_BY_NAME


def role_requires_manifest_resolution(
    role: str,
    *,
    agents_config: AgentsConfig | None = None,
) -> bool:
    role_config = agents_config.for_role(role) if agents_config is not None else None
    return not is_built_in_profile_name(profile_name_for_role(role, role_config))


def profile_from_role_config(
    role: str,
    role_config: RoleAgentConfig | None,
    *,
    available_profiles: Mapping[str, AgentProfile] | None = None,
    manifest_definitions: Sequence["LoadedAgentDefinition"] = (),
) -> AgentProfile:
    base_profile_name = profile_name_for_role(role, role_config)
    catalog = (
        dict(available_profiles)
        if available_profiles is not None
        else available_profile_catalog(manifest_definitions=manifest_definitions)
    )
    base_profile = catalog.get(base_profile_name)
    if base_profile is None:
        raise ValueError(
            f"Role {role!r} maps to profile {base_profile_name!r}, "
            "but no effective profile with that name is available."
        )

    effective_profile = base_profile
    if _role_override_present(role_config):
        assert role_config is not None
        effective_profile = replace(
            base_profile,
            source=CONFIGURED_PROFILE_SOURCE,
            cli_override=role_config.cli if role_config.cli is not None else base_profile.cli_override,
            model_override=role_config.model if role_config.model is not None else base_profile.model_override,
            permission_policy=merge_permission_policy(
                base_profile.permission_policy,
                role_config.permission_policy,
            ),
        )

    runtime_metadata = _runtime_resolution_metadata(
        role=role,
        base_profile=base_profile,
        role_config=role_config,
        manifest_metadata=_manifest_metadata_by_profile_name(manifest_definitions).get(
            base_profile.name
        ),
    )
    metadata = dict(effective_profile.metadata)
    metadata[PROFILE_RUNTIME_METADATA_KEY] = runtime_metadata
    return replace(effective_profile, metadata=metadata)


def available_profile_catalog(
    *,
    manifest_definitions: Sequence["LoadedAgentDefinition"] = (),
) -> dict[str, AgentProfile]:
    profiles = {profile.name: profile for profile in built_in_profiles()}
    for definition in manifest_definitions:
        if definition.name in profiles:
            raise ValueError(
                "Manifest-backed profile "
                f"{definition.name!r} conflicts with an existing built-in profile name."
            )
        profiles[definition.name] = definition.to_profile()
    return profiles


def _manifest_metadata_by_profile_name(
    manifest_definitions: Sequence["LoadedAgentDefinition"],
) -> dict[str, dict[str, str]]:
    return {
        definition.name: {
            "manifest_scope": definition.manifest_scope,
            "manifest_path": str(definition.manifest_path),
        }
        for definition in manifest_definitions
    }


def _runtime_resolution_metadata(
    *,
    role: str,
    base_profile: AgentProfile,
    role_config: RoleAgentConfig | None,
    manifest_metadata: Mapping[str, str] | None,
) -> dict[str, object]:
    return {
        "runtime_role": role,
        "selected_profile_name": base_profile.name,
        "selected_profile_source": base_profile.source,
        "selected_via_role_config": (
            role_config.profile is not None if role_config is not None else False
        ),
        "role_overrides": {
            "cli": role_config.cli is not None if role_config is not None else False,
            "model": role_config.model is not None if role_config is not None else False,
            "permission_policy": (
                role_config.permission_policy is not None
                if role_config is not None
                else False
            ),
        },
        **(dict(manifest_metadata) if manifest_metadata is not None else {}),
    }


def resolve_agent_profile(
    role: str,
    *,
    agents_config: AgentsConfig | None = None,
    normalized_profiles: Mapping[str, AgentProfile] | None = None,
    manifest_definitions: Sequence["LoadedAgentDefinition"] = (),
) -> AgentProfile:
    return resolve_runtime_role_profile(
        role,
        agents_config=agents_config,
        normalized_profiles=normalized_profiles,
        manifest_definitions=manifest_definitions,
    )


def normalize_agent_profiles(
    *,
    agents_config: AgentsConfig | None = None,
    manifest_definitions: Sequence["LoadedAgentDefinition"] = (),
) -> dict[str, AgentProfile]:
    catalog = available_profile_catalog(manifest_definitions=manifest_definitions)
    profiles: dict[str, AgentProfile] = {}
    for role in ROLE_NAMES:
        role_config = agents_config.for_role(role) if agents_config is not None else None
        profiles[role] = profile_from_role_config(
            role,
            role_config,
            available_profiles=catalog,
            manifest_definitions=manifest_definitions,
        )
    return profiles


def resolve_runtime_role_profile(
    role: str,
    *,
    agents_config: AgentsConfig | None = None,
    normalized_profiles: Mapping[str, AgentProfile] | None = None,
    manifest_definitions: Sequence["LoadedAgentDefinition"] = (),
) -> AgentProfile:
    if role not in ROLE_TO_PROFILE_NAME:
        raise ValueError(f"Unknown role: {role!r}. Valid roles: {ROLE_NAMES}")

    if normalized_profiles is not None:
        profile = normalized_profiles.get(role)
        if profile is None:
            role_config = agents_config.for_role(role) if agents_config is not None else None
            profile_name = profile_name_for_role(role, role_config)
            raise ValueError(
                f"Role {role!r} maps to profile {profile_name!r}, "
                "but no effective profile with that name is available."
            )
        return profile

    role_config = agents_config.for_role(role) if agents_config is not None else None
    return profile_from_role_config(
        role,
        role_config,
        manifest_definitions=manifest_definitions,
    )

File: runtime/src/agent/profiles.ts
import path from "node:path";
import os from "node:os";

import {
  defaultAgentPermissionPolicy,
  mergePermissionPolicy,
  parsePermissionPolicyOverride,
  type AgentPermissionPolicy,
  type AgentPermissionPolicyOverride,
  type WorktreePermissionPolicy
} from "./permissions.js";

export const BUILTIN_PROFILE_SOURCE = "built_in";
export const CONFIGURED_PROFILE_SOURCE = "configured";
export const PROJECT_PROFILE_SOURCE = "project";
export const USER_PROFILE_SOURCE = "user";
export const PROFILE_RUNTIME_METADATA_KEY = "dormammu_runtime";

export type RoleTaxonomyEntry = {
  name: string;
  scope: string;
  description: string;
};

export const ROLE_TAXONOMY = [
  {
    name: "refiner",
    scope: "runtime",
    description: "Refines the raw request into explicit implementation requirements."
  },
  {
    name: "analyzer",
    scope: "goals_autonomous_only",
    description:
      "Analyzes scheduled goals or autonomous repository context before a runtime prompt is queued."
  },
  {
    name: "planner",
    scope: "runtime_and_goals_prelude",
    description: "Plans the task and updates the operator-facing workflow state."
  },
  {
    name: "designer",
    scope: "goals_prelude_only",
    description:
      "Adds optional technical design context to goals-scheduler prompts; runtime review reads its design document when present."
  },
  {
    name: "developer",
    scope: "runtime",
    description: "Implements the active product-code slice under supervisor control."
  },
  {
    name: "tester",
    scope: "runtime",
    description: "Runs black-box validation against the observable behavior of the slice."
  },
  {
    name: "reviewer",
    scope: "runtime",
    description: "Reviews changed code for regressions, bugs, and missing coverage."
  },
  {
    name: "committer",
    scope: "runtime",
    description: "Prepares validated changes for version-control handoff."
  },
  {
    name: "evaluator",
    scope: "goals_checkpoint_only",
    description: "Evaluates goals-scheduler plan checkpoints and final goal completion."
  }
] as const satisfies readonly RoleTaxonomyEntry[];

export type RoleName = (typeof ROLE_TAXONOMY)[number]["name"];
export const ROLE_NAMES = ROLE_TAXONOMY.map((entry) => entry.name) as RoleName[];
export const RUNTIME_PIPELINE_ROLE_NAMES = [
  "refiner",
  "planner",
  "developer",
  "tester",
  "reviewer",
  "committer"
] as const;
export const GOALS_PRELUDE_ROLE_NAMES = ["analyzer", "planner", "designer"] as const;
export const GOALS_OR_AUTONOMOUS_ONLY_ROLE_NAMES = [
  "analyzer",
  "designer",
  "evaluator"
] as const;

export type AgentProfile = {
  name: string;
  description: string;
  source: string;
  prompt_body: string | null;
  cli_override: string | null;
  model_override: string | null;
  permission_policy: AgentPermissionPolicy;
  preloaded_skills: string[];
  metadata: Record<string, unknown>;
};

export type RoleAgentConfig = {
  profile: string | null;
  cli: string | null;
  model: string | null;
  permission_policy: AgentPermissionPolicyOverride | null;
};

export type AgentsConfig = Record<RoleName, RoleAgentConfig>;

export function defaultRoleAgentConfig(): RoleAgentConfig {
  return {
    profile: null,
    cli: null,
    model: null,
    permission_policy: null
  };
}

export function defaultAgentsConfig(): AgentsConfig {
  return Object.fromEntries(
    ROLE_NAMES.map((role) => [role, defaultRoleAgentConfig()])
  ) as AgentsConfig;
}

export function builtInProfiles(): AgentProfile[] {
  return ROLE_TAXONOMY.map((entry) => ({
    name: entry.name,
    description: entry.description,
    source: BUILTIN_PROFILE_SOURCE,
    prompt_body: null,
    cli_override: null,
    model_override: null,
    permission_policy: defaultAgentPermissionPolicy(),
    preloaded_skills: [],
    metadata: {}
  }));
}

export function profileNameForRole(
  role: string,
  roleConfig: RoleAgentConfig | null = null
): string {
  assertRoleName(role);
  return roleConfig?.profile ?? role;
}

export function builtInProfileForRole(role: string): AgentProfile {
  const profileName = profileNameForRole(role);
  const profile = builtInProfiles().find((candidate) => candidate.name === profileName);
  if (!profile) {
    throw new Error(
      `Role ${JSON.stringify(role)} maps to profile ${JSON.stringify(profileName)}, but no built-in profile with that name is available.`
    );
  }
  return profile;
}

export function availableProfileCatalog(
  manifestProfiles: readonly AgentProfile[] = []
): Record<string, AgentProfile> {
  const profiles: Record<string, AgentProfile> = {};
  for (const profile of builtInProfiles()) {
    profiles[profile.name] = cloneAgentProfile(profile);
  }
  for (const profile of manifestProfiles) {
    if (profiles[profile.name]) {
      throw new Error(
        `Manifest-backed profile ${JSON.stringify(profile.name)} conflicts with an existing built-in profile name.`
      );
    }
    profiles[profile.name] = cloneAgentProfile(profile);
  }
  return profiles;
}

export function profileFromRoleConfig(
  role: string,
  roleConfig: RoleAgentConfig | null = null,
  options: {
    availableProfiles?: Record<string, AgentProfile> | null;
    manifestProfiles?: readonly AgentProfile[];
  } = {}
): AgentProfile {
  const baseProfileName = profileNameForRole(role, roleConfig);
  const catalog =
    options.availableProfiles ?? availableProfileCatalog(options.manifestProfiles ?? []);
  const baseProfile = catalog[baseProfileName];
  if (!baseProfile) {
    throw new Error(
      `Role ${JSON.stringify(role)} maps to profile ${JSON.stringify(baseProfileName)}, but no effective profile with that name is available.`
    );
  }
  const roleOverridePresent = roleAgentOverridePresent(roleConfig);
  const effectiveProfile = roleOverridePresent
    ? {
        ...cloneAgentProfile(baseProfile),
        source: CONFIGURED_PROFILE_SOURCE,
        cli_override: roleConfig?.cli ?? baseProfile.cli_override,
        model_override: roleConfig?.model ?? baseProfile.model_override,
        permission_policy: mergePermissionPolicy(
          baseProfile.permission_policy,
          roleConfig?.permission_policy
        )
      }
    : cloneAgentProfile(baseProfile);
  return {
    ...effectiveProfile,
    metadata: {
      ...effectiveProfile.metadata,
      [PROFILE_RUNTIME_METADATA_KEY]: runtimeResolutionMetadata({
        role,
        baseProfile,
        roleConfig
      })
    }
  };
}

export function normalizeAgentProfiles(
  options: {
    agentsConfig?: AgentsConfig | null;
    manifestProfiles?: readonly AgentProfile[];
  } = {}
): Record<RoleName, AgentProfile> {
  const catalog = availableProfileCatalog(options.manifestProfiles ?? []);
  return Object.fromEntries(
    ROLE_NAMES.map((role) => [
      role,
      profileFromRoleConfig(role, options.agentsConfig?.[role] ?? null, {
        availableProfiles: catalog,
        manifestProfiles: options.manifestProfiles ?? []
      })
    ])
  ) as Record<RoleName, AgentProfile>;
}

export function resolveRuntimeRoleProfile(
  role: string,
  options: {
    agentsConfig?: AgentsConfig | null;
    normalizedProfiles?: Partial<Record<RoleName, AgentProfile>> | null;
    manifestProfiles?: readonly AgentProfile[];
  } = {}
): AgentProfile {
  assertRoleName(role);
  const roleName = role as RoleName;
  if (options.normalizedProfiles) {
    const profile = options.normalizedProfiles[roleName];
    if (!profile) {
      const roleConfig = options.agentsConfig?.[roleName] ?? null;
      const profileName = profileNameForRole(role, roleConfig);
      throw new Error(
        `Role ${JSON.stringify(role)} maps to profile ${JSON.stringify(profileName)}, but no effective profile with that name is available.`
      );
    }
    return cloneAgentProfile(profile);
  }
  return profileFromRoleConfig(role, options.agentsConfig?.[roleName] ?? null, {
    manifestProfiles: options.manifestProfiles ?? []
  });
}

export function parseAgentsConfig(
  value: unknown,
  options: { configPath?: string | null } = {}
): AgentsConfig | null {
  if (value === undefined || value === null) {
    return null;
  }
  const source = options.configPath ?? "dormammu.json";
  const payload = coerceRecord(value, "agents", source);
  return Object.fromEntries(
    ROLE_NAMES.map((role) => [
      role,
      parseRoleAgentConfig(payload[role], {
        role,
        configPath: options.configPath ?? null
      })
    ])
  ) as AgentsConfig;
}

export function parseRoleAgentConfig(
  value: unknown,
  options: { role: RoleName; configPath?: string | null }
): RoleAgentConfig {
  if (value === undefined || value === null) {
    return defaultRoleAgentConfig();
  }
  const source = options.configPath ?? "dormammu.json";
  const payload = coerceRecord(value, `agents.${options.role}`, source);
  return {
    profile: parseOptionalString(payload.profile, `agents.${options.role}.profile`, source),
    cli: parseCli(payload.cli, options.configPath ?? null, `agents.${options.role}.cli`, source),
    model: parseOptionalString(payload.model, `agents.${options.role}.model`, source),
    permission_policy:
      payload.permission_policy === undefined || payload.permission_policy === null
        ? null
        : parsePermissionPolicyOverride(payload.permission_policy, {
            configRoot: options.configPath ? path.dirname(options.configPath) : null,
            fieldName: `agents.${options.role}.permission_policy`,
            source
          })
  };
}

export function mergeRoleAgentConfig(
  base: RoleAgentConfig,
  override: RoleAgentConfig
): RoleAgentConfig {
  return {
    profile: override.profile ?? base.profile,
    cli: override.cli ?? base.cli,
    model: override.model ?? base.model,
    permission_policy: mergePermissionPolicyOverride(base.permission_policy, override.permission_policy)
  };
}

export function mergeAgentsConfig(
  base: AgentsConfig | null,
  override: AgentsConfig | null
): AgentsConfig | null {
  if (!base) {
    return override;
  }
  if (!override) {
    return base;
  }
  return Object.fromEntries(
    ROLE_NAMES.map((role) => [role, mergeRoleAgentConfig(base[role], override[role])])
  ) as AgentsConfig;
}

export function resolveProfileCli(
  profile: AgentProfile,
  activeAgentCli: string | null
): string | null {
  return profile.cli_override ?? activeAgentCli;
}

export function profileWorktreePolicy(profile: AgentProfile): WorktreePermissionPolicy {
  return profile.permission_policy.worktree;
}

function mergePermissionPolicyOverride(
  base: AgentPermissionPolicyOverride | null,
  override: AgentPermissionPolicyOverride | null
): AgentPermissionPolicyOverride | null {
  if (!base || permissionPolicyOverrideIsEmpty(base)) {
    return override;
  }
  if (!override || permissionPolicyOverrideIsEmpty(override)) {
    return base;
  }
  return {
    tools: mergePolicyOverrideBlock(base.tools, override.tools),
    skills: mergePolicyOverrideBlock(base.skills, override.skills),
    filesystem: mergePolicyOverrideBlock(base.filesystem, override.filesystem),
    network: mergePolicyOverrideBlock(base.network, override.network),
    worktree: mergePolicyOverrideBlock(base.worktree, override.worktree)
  };
}

function mergePolicyOverrideBlock<T extends { default?: string | null; rules: unknown[] }>(
  base: T | null | undefined,
  override: T | null | undefined
): T | null {
  if (!base) {
    return override ?? null;
  }
  if (!override) {
    return base;
  }
  return {
    default: override.default ?? base.default ?? null,
    rules: [...base.rules, ...override.rules]
  } as T;
}

function runtimeResolutionMetadata(options: {
  role: string;
  baseProfile: AgentProfile;
  roleConfig: RoleAgentConfig | null;
}): Record<string, unknown> {
  const roleConfig = options.roleConfig;
  return {
    runtime_role: options.role,
    selected_profile_name: options.baseProfile.name,
    selected_profile_source: options.baseProfile.source,
    selected_via_role_config: roleConfig?.profile !== null && roleConfig?.profile !== undefined,
    role_overrides: {
      cli: roleConfig?.cli !== null && roleConfig?.cli !== undefined,
      model: roleConfig?.model !== null && roleConfig?.model !== undefined,
      permission_policy:
        roleConfig?.permission_policy !== null && roleConfig?.permission_policy !== undefined
    }
  };
}

function roleAgentOverridePresent(roleConfig: RoleAgentConfig | null): boolean {
  return Boolean(roleConfig?.cli ?? roleConfig?.model ?? roleConfig?.permission_policy);
}

function permissionPolicyOverrideIsEmpty(override: AgentPermissionPolicyOverride): boolean {
  return (
    !override.tools &&
    !override.skills &&
    !override.filesystem &&
    !override.network &&
    !override.worktree
  );
}

function cloneAgentProfile(profile: AgentProfile): AgentProfile {
  return {
    ...profile,
    permission_policy: {
      tools: { ...profile.permission_policy.tools, rules: [...profile.permission_policy.tools.rules] },
      skills: { ...profile.permission_policy.skills, rules: [...profile.permission_policy.skills.rules] },
      filesystem: {
        ...profile.permission_policy.filesystem,
        rules: [...profile.permission_policy.filesystem.rules]
      },
      network: { ...profile.permission_policy.network, rules: [...profile.permission_policy.network.rules] },
      worktree: { ...profile.permission_policy.worktree, rules: [...profile.permission_policy.worktree.rules] }
    },
    preloaded_skills: [...profile.preloaded_skills],
    metadata: { ...profile.metadata }
  };
}

function parseCli(
  value: unknown,
  configPath: string | null,
  fieldName: string,
  source: string
): string | null {
  const raw = parseOptionalString(value, fieldName, source);
  if (raw === null) {
    return null;
  }
  const candidate = expandUserPath(raw);
  const configDir = configPath ? path.dirname(configPath) : null;
  if (path.isAbsolute(candidate)) {
    return candidate;
  }
  if (configDir && (raw.includes("/") || raw.startsWith("."))) {
    return path.resolve(configDir, candidate);
  }
  return candidate;
}

function parseOptionalString(value: unknown, fieldName: string, source: string): string | null {
  if (value === undefined || value === null) {
    return null;
  }
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${fieldName} must be a non-empty string in ${source}`);
  }
  return value.trim();
}

function assertRoleName(role: string): asserts role is RoleName {
  if (!ROLE_NAMES.includes(role as RoleName)) {
    throw new Error(`Unknown role: ${JSON.stringify(role)}. Valid roles: ${ROLE_NAMES.join(",")}`);
  }
}

function coerceRecord(value: unknown, fieldName: string, source: string): Record<string, unknown> {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw new Error(`${fieldName} must be a JSON object in ${source}`);
  }
  return value as Record<string, unknown>;
}

function expandUserPath(value: string): string {
  if (value === "~") {
    return os.homedir();
  }
  if (value.startsWith(`~${path.sep}`) || value.startsWith("~/")) {
    return path.join(os.homedir(), value.slice(2));
  }
  return value;
}
